server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java

```diff
@@ -368,6 +372,45 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
         builder.endObject();
     }
 
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        QueryBuilder rewritten = super.doRewrite(queryRewriteContext);
+
+        if (rewritten == this && queryRewriteContext.getClass() == QueryRewriteContext.class) {
+            ResolvedIndices resolvedIndices = queryRewriteContext.getResolvedIndices();
+            if (resolvedIndices != null) {
+                Collection<IndexMetadata> indexMetadataCollection = resolvedIndices.getConcreteLocalIndicesMetadata().values();
+
+                boolean foundNonInferenceField = false;
+                String inferenceFieldQueryName = null;
+                for (IndexMetadata indexMetadata : indexMetadataCollection) {
+                    InferenceFieldMetadata inferenceFieldMetadata = indexMetadata.getInferenceFields().get(fieldName);
+                    if (inferenceFieldMetadata != null) {
+                        if (inferenceFieldQueryName != null
+                            && inferenceFieldMetadata.getQueryName().equals(inferenceFieldQueryName) == false) {
+                            throw new IllegalArgumentException("Detected incompatible inference field queries");
```

Comment:
These error messages leave much to be desired. Will iterate on them to make them more informative.